tests/numerics/test_time_step_control.py

```diff
@@ -348,6 +348,30 @@ def test_number_of_recomp_attempts_greater_than_zero(
             )
         assert msg in str(excinfo.value)
 
+    @pytest.mark.parametrize("num_time_steps", [10, 20])
```

Comment:
@jwboth Regarding your question about the time it takes to run the test: Would it work to say, e.g. `[3, 5]` here? Or `[10, 11]`?

I'm not very worried about the cost of this test, but if we can reduce it further, that would be great.

Reply:
Now the test is parametrized with:
-  number of time steps [1, 3]
-  final time [1.0e-14,1.0e0,1.0e+14]

Reply:
Not sure if this is relevant: @OmarDuran, did you test whether one of the options would have still broken the test before? Not every value for `num_time_steps` has led to the unexpected behavior of `TimeManager`. 

I like the idea of testing small final times to address the tolerance question.

Reply:
@jwboth with the implementation you did in `def final_time_reached(self)` the code works correctly for any `num_time_steps`. However the question that remains was to test values very close to the final time.

Let set `a = self.time` and `b = self.time_final`. To mimic the situation `a` very close to `b`, it is enough to test very small values of `a` and `b` in the sense of `absolute(a - b) <= (atol + rtol * absolute(b))`. This cover all the cases for regular time steps discretizations. I believe possible situations associated with irregular time steps belong to another issue and a future pull request.

Reply:
@OmarDuran: What do you mean by regular time step? To me, this is a highly problematic concept, for the following reason:

In addition to running simulation on the unit interval, I regularly consider cases where the simulation first runs for a couple of hundred years to reach a steady state, followed by a period where injection rates is changed on a schedule of, say, minutes. On the other hand, if we were to do simulations inspired by the FluidFlower CO2 experiments, the time step size needed in OPM are commonly down to fractions of milliseconds, and PorePy would likely need even smaller ones. Unless we make the code robust in such wide ranges, I don't think we can consider the issue closed.

Reply:
@keileg: I believe this  PR do fix the issue #939. In this case, it is not about running simulations on a unit interval nor dealing with different time step sizes in the same simulation, but rather to ensure that the number of time steps performed is the one requested. This works perfectly for regular time step sizes (constant time step size) and large simulation times. You have elaborated on an issue of different nature when it comes to irregular time step sizes (no constant time steps). Considering the case you describe above, I believe we should create another issue specific to this case.

Reply:
Hereby, after a quick local test, I can assure that the values chosen by @OmarDuran would not break the initial test given the old implementation of checking the final state of the simulation. Also values [3,5] for num_time_steps pass, even for final time math.pi.  Finally, the combination of 7 time steps and final time 1e-14 breaks the test. My suggestions would be to aim at four tests with 2 breaking the previous implementation - one per final time. Thus, althoug I see the critcism around final_time=1, I would suggest the following selection:
   `@pytest.mark.parametrize(
        "num_time_steps, final_time",
        [
            (3, 1.0e-14),
            (7, 1.0e-14),
            (3, pi),
            (10, 1.),
        ],
    )`
Here, the original implementation would break for the second and fourth test. Repalcing pi with 1 in the case (3, pi) would be sensible to be as well - no strong opinion here.

Reply:
Let's do that.

Reply:
@keileg Done. I flipped the coin and chose not to include pi via an extra import.